### ISSUE TITLE
Yield batches from ogr_read_arrow

### ISSUE THREAD

**Issue by kylebarron** [geopandas/pyogrio#205]
I'm playing around more with reading arrow tables from pyogrio and it's really exciting. It does feel like having some API to yield batches would be helpful to work with larger datasets. @jorisvandenbossche wrote in https://github.com/geopandas/pyogrio/pull/155:

> Returning a materialized Table (from `read_all()`) is fine for our current use case, but I can imagine that in the future, we might want to expose an iterative RecordBatchReader as well (eg as batch-wise input to query engine).
> When we want to do that, I assume that we somehow need to keep the GDAL Dataset alive (putting it in a Python object (wrapping in a small class, or putting in a PyCapsule with destructor), and keeping a reference to that object from the RecordBatchReader).

I've never touched ogr bindings before, but naively it seems the easiest way to do this is by using a context manager:
```py
with open_arrow("file.shp") as reader:
    for record_batch in reader:
		table
```
would that work? just putting a yield [here](https://github.com/geopandas/pyogrio/pull/155/files#diff-772414414b26a9df2e74063fd9e20d64cbc189d0e7e5671613c3bcc39faf0e47R1071-R1081)?

**Comment by brendan-ward** [comment]
Do you envision that there would be a counterpart function to write data in batches, via the Arrow I/O API (once available)?

**Comment by kylebarron** [comment]
That seems entirely dependent on GDAL? https://gdal.org/development/rfc/rfc86_column_oriented_api.html says

> Potential future work, not in the scope of this RFC, could be the addition of a column-oriented method to write new features, a WriteRecordBatch() method.

I would of course love for that to be added to GDAL, and getting greater adoption of RFC 86 seems very helpful for that.

https://github.com/geopandas/pyogrio/pull/206 appeared to work on my local machine 🤷‍♂️. If a maintainer could enable CI on that PR that would be helpful!

**Comment by kylebarron** [comment]
If GDAL were to add it, I think a similar API like

```py
with write_ogr_batches("file.gpkg", arrow_schema) as writer:
    writer.write_batch(batch)
```

could make sense. But for my own needs I think I'm more likely to only _write_ to geoparquet and thus not use OGR as much for writing

**Comment by jorisvandenbossche** [comment]
@kylebarron thanks for looking into this! That seems like a really nice idea, I didn't think about a contextmanager to keep the dataset alive.

**Comment by jorisvandenbossche** [comment]
This was closed by #206
